### PR TITLE
Server heartbeats

### LINKED ISSUE #343: Server heartbeat
Implement a server heartbeat that periodically sends a message with usage data, which right now is number of rows read and used storage. This is useful for monitoring the databases but also for usage tracking.

### PR DESCRIPTION
**Pull request by penberg**
This adds support for server heartbeats.

The patch adds three new command line options:

  `--heartbeat-url URL` is the URL to send a HTTP POST request to
  periodically to indicate server heatbeat.

  `--heartbeat-auth AUTH` is the HTTP "Authorization" header to send
  as part of the HTTP POST request.

  `--heartbeat-period SECS` is the heartbeat period in seconds.

The payload of the HTTP POST request is pretty simple for now:

```json
{"rows_read":1234,"rows_written":8765}
```

Fixes #343

### CODE REVIEW

File: sqld/src/main.rs [hunk 1]
```diff
@@ -133,6 +133,22 @@ struct Cli {
...
+
+    /// The URL to send a server heartbeat `POST` request to.
+    /// By default, the server doesn't send a heartbeat.
+    #[clap(long, env = "SQLD_HEARTBEAT_URL")]
+    heartbeat_url: Option<String>,
+
+    /// The HTTP "Authornization" header to include in the a server heartbeat
+    /// `POST` request.
+    /// By default, the server doesn't send a heartbeat.
+    #[clap(long, env = "SQLD_HEARTBEAT_URL")]
```
Comment (MarinPostma) [inline sqld/src/main.rs:145]: you can add a `require` to create a dependency with other arguments: https://docs.rs/clap/latest/clap/struct.Arg.html#method.requires

File: sqld/src/main.rs [hunk 2]
```diff
@@ -133,6 +133,22 @@ struct Cli {
...
+    heartbeat_auth: Option<String>,
+
+    /// The heartbeat time period in seconds.
+    /// By default, the server doesn't send a heartbeat.
+    #[clap(long, env = "SQLD_HEARTBEAT_PERIOD_S")]
+    heartbeat_period_s: Option<u64>,
```
Comment (MarinPostma) [inline sqld/src/main.rs:151]: same as auth, and maybe set a sensible default

File: sqld/src/main.rs (hunk 1 above)
Comment (MarinPostma) [inline sqld/src/main.rs:145]:
```suggestion
    #[clap(long, env = "SQLD_HEARTBEAT_AUTH")]
```

File: sqld/src/lib.rs [hunk 3]
```diff
@@ -145,6 +149,38 @@ async fn run_service(
         });
     }
 
+    match (&config.heartbeat_url, &config.heartbeat_period) {
+        (Some(heartbeat_url), Some(heartbeat_period)) => {
+            tracing::info!(
+                "Server sending heartbeat to URL {} every {}",
+                heartbeat_url,
+                humantime::format_duration(*heartbeat_period).to_string()
```
Comment (MarinPostma) [inline sqld/src/lib.rs:157]: you can use the `Debug` implementation of Duration to get nice formatting: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8eca3df4fb1a508658d91db44eb35174

no need for humantime

**Review by MarinPostma — CHANGES_REQUESTED** [review]

File: sqld/src/heartbeat.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    update_period: Duration,
+    stats: Stats,
+) -> Result<()> {
+    let client = reqwest::Client::new();
+    loop {
+        sleep(update_period).await;
+        let body = serde_json::json!({
+            "rows_read": stats.rows_read(),
+            "rows_written": stats.rows_written(),
+        });
```
[the remark this answers is not in the document]
Comment (penberg) [reply sqld/src/heartbeat.rs]: You mentioned privately that you actually meant `StatsResponse`, so switched over to that.

File: sqld/src/heartbeat.rs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+use crate::http::stats::StatsResponse;
+use crate::stats::Stats;
+
+pub async fn server_heartbeat(
+    url: String,
+    auth: Option<String>,
+    update_period: Duration,
+    stats: Stats,
+) -> Result<()> {
```
Comment (MarinPostma) [inline sqld/src/heartbeat.rs:13]: this method does not return an error anymore

**Review by MarinPostma — APPROVED** [review]

**Pull request merged**